addons/udes_stock/models/stock_picking_batch.py

```diff
@@ -298,63 +303,128 @@ def get_next_task(self, skipped_product_ids=None, task_grouping_criteria=None):
-            # ends up in the batch::next response!
-            # HERE: this will break in case we cannot guarantee that all
-            # the move lines of the task belong to the same picking
-            task.update(task_mls._prepare_task_info())
+    def _populate_next_task(self, move_lines, task_grouping_criteria, priority_ml=False):
+        """Populate the next task from the available move lines and grouping.
 
-            if task_mls[0].picking_id.picking_type_id.u_user_scans == "product":
-                # NB: adding 1 to consider the group removed by next()
-                task["num_tasks_to_pick"] = len(list(grouped_mls)) + 1
-                task["move_line_ids"] = task_mls.ids
-            else:
-                # TODO: check pallets of packages if necessary
-                task["num_tasks_to_pick"] = len(todo_mls.mapped("package_id"))
-                task["move_line_ids"] = todo_mls.filtered(
-                    lambda ml: ml.package_id == task_mls[0].package_id
-                ).ids
+        Optionally specify a priority moveline to be in the next task.
+        """
+        task = {
+            "num_tasks_to_pick": 0,
+            "move_line_ids": [],
+            "confirmations": [],
+        }
+        if not move_lines:
+            return task
+
+        if task_grouping_criteria is None:
+            task_grouping_criteria = self._get_task_grouping_criteria()
+
+        grouped_mls = move_lines.groupby(task_grouping_criteria)
+        _key, task_mls = next(grouped_mls)
+        # Iterate through grouped_mls until we find the group with the
+        # priority move line in it
+        if priority_ml:
+            while priority_ml not in task_mls:
+                _key, task_mls = next(grouped_mls)
+        num_mls = len(task_mls)
+        pick_seq = task_mls[0].picking_id.sequence
+        _logger.debug(_("Batch '%s': creating a task for %s move line%s; "
+                        "the picking sequence of the first move line is %s"),
+                      self.name,
+                      num_mls,
+                      "" if num_mls == 1 else "s",
+                      pick_seq if pick_seq is not False else "not determined")
+
+        # NB: adding all the MLs state to the task; this is what
+        # ends up in the batch::next response!
+        # HERE: this will break in case we cannot guarantee that all
+        # the move lines of the task belong to the same picking
+        task.update(task_mls._prepare_task_info())
+
+        if task_mls[0].picking_id.picking_type_id.u_user_scans == "product":
+            # NB: adding 1 to consider the group removed by next()
+            task["num_tasks_to_pick"] = len(list(grouped_mls)) + 1
+            task["move_line_ids"] = task_mls.ids
         else:
-            _logger.debug(
-                _("Batch '%s': no available move lines for creating " "a task"), self.name
-            )
+            # TODO: check pallets of packages if necessary
```

Comment:
Do we still want to keep TODO?

Reply:
This TODO was already in the code, it was just moved here. I haven't addressed it so I think it still stands